Add MPU support

pengi opened the pull request:
Add MPU support, based on the code and interface of SCB implememation.

pengi committed:
Add MPU support

Currently only for ARMv8, since it seems to be implementation dependent
on other architectures. May be updated later

pengi commented:
Since @swordligit are working with MPU, could you check if you have any thoughts? or if I'm missing something. I'm only checking my Cortex-M33 in the initial state, with all MPU disabled.

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+def get_mpu_common_regs(model):
+    def mair_attr_fields(num, offset):
+        return [
+            FieldBitfieldEnum(f"Outer {num}", offset+4, 4, [
```
space around +

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+                (0b1101, None, "Normal Memory, Inner Write-Back Non-transient, Allocate W", None),
+                (0b1110, None, "Normal Memory, Inner Write-Back Non-transient, Allocate R", None),
+                (0b1111, None, "Normal Memory, Inner Write-Back Non-transient, Allocate RW", None),
+            ], "Device attributes. Specifies the memory attributes for Device. (Valid only if Outer == 0)")
```
Normal attributes. Specifies the memory attributes for Normal. (Valid only if Outer != 0)

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+def get_mpu_region_regs(model):
+    return [
+        RegisterDef(f"MPU_RBAR", f"MPU Region Base Address Register", 0xE000ED9C, 4, [
+            FieldBitfield("BASE", 5, 27,
```
Now the base shows like:

MPU_RBAR                         = 00500006 
    BASE                           0050000. - 0028000

in fact the address is still 0x50000,  shows 0x28000 maybe confused.

pengi replied:
Oh right, good catch. Got to set that as a Map field then

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+                        "Base address. Contains bits [31:5] of the lower inclusive limit of the selected MPU memory region"),
+            FieldBitfieldEnum("SH", 3, 2, [
+                            (0b00, False, "Non-shareable.", None),
+                            (0b01, False, "invalid", None),
```
Invalid

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+        ]),
+        RegisterDef(f"MPU_RLAR", f"MPU Region Limit Address Register", 0xE000EDA0, 4, [
+            FieldBitfield("LIMIT", 5, 27,
+                        "Limit address. Contains bits [31:5] of the upper inclusive limit of the selected MPU memory region"),
```
like the MPU_RBAR, if shows real limit address may better.

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -0,0 +1,248 @@
+                "4100c270": ["M7", "v7"],
+                # TODO: support ARMv8-M, for now pretend it's v7, since it's similar
+                "4100d200": ["M23", "v8"],
+                "4100d210": ["M33", "v8"],
```
My cpuid is 
CPUID                            = 630f1321 
    Implementer                    63...... - 63             
    Variant                        ..0..... - Revision: r0pX 
    Architecture                   ...f.... - ARMv7-M        
    PartNo                         ....132. - 132            
    Revision                       .......1 - Patch: rXp1
but it is cortex-m33, and armv8m. please and this to dict?
                "63001320": ["M33", "v8"]

pengi replied:
Gosh... CPUID format is a mess. Apparently Architecture field of 1111 is the architecture in ARMv7-M reference manual, but in ARMv8-M it's also a valid value... Need to investigate a bit.

But what CPU are you using? Since "Implementer" 63 is not ARM, so I guess it's a clone? ARM should be 41

pengi replied:
I've added it for now at least

swordligit replied:
Thank you, it is China star-mc1, a clone of cortex-m33.

pengi committed:
MPU fixes

pengi committed:
MPU: Add conditions to fields

Some fields change format depending on other bits in the register.

Add support for conditional fields

pengi commented:
I also added support for conditional fields, so Device or Inner is correctly filtered, hopefully.

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -55,32 +55,34 @@ def mair_attr_fields(num, offset):
+                    (0b1101, None, "Normal Memory, Inner Write-Back Non-transient, Allocate W", None),
+                    (0b1110, None, "Normal Memory, Inner Write-Back Non-transient, Allocate R", None),
+                    (0b1111, None, "Normal Memory, Inner Write-Back Non-transient, Allocate RW", None),
+                ], "Device attributes. Specifies the memory attributes for Device.")
```
Normal attributes. Specifies the memory attributes for Normal.

swordligit commented on arm_gdb/mpu.py:
```diff
@@ -215,6 +215,7 @@ def invoke(self, argument, from_tty):
                 # TODO: support ARMv8-M, for now pretend it's v7, since it's similar
                 "4100d200": ["M23", "v8"],
                 "4100d210": ["M33", "v8"],
+                "63001320": ["M55", "v8"],
```
So just use ["Star", "v8"]?

swordligit commented:
I have test on my board, it works pefect, thanks a lot.

pengi commented:
Thanks for helping out. I'll merge then :)

pengi commented:
New version is out too. 0.9.10 should include it. Just update via pip :)

swordligit commented:
sorry, not a big issue, there is a mistake.
```
            fieldconditional(lambda value: value & (0xf0 << offset) != 0,
                fieldbitfieldenum(f"inner {num}", offset + 0, 4, [
                    (0b0000, none, "unpredictable", none),
                    (0b0001, none, "normal memory, inner write-through transient, allocate w", none),
                    (0b0010, none, "normal memory, inner write-through transient, allocate r", none),
                    (0b0011, none, "normal memory, inner write-through transient, allocate rw", none),
                    (0b0100, none, "normal memory, inner non-cacheable", none),
                    (0b0101, none, "normal memory, inner write-back transient, allocate w", none),
                    (0b0110, none, "normal memory, inner write-back transient, allocate r", none),
                    (0b0111, none, "normal memory, inner write-back transient, allocate rw", none),
                    (0b1000, none, "normal memory, inner write-through non-transient", none),
                    (0b1001, none, "normal memory, inner write-through non-transient, allocate w", none),
                    (0b1010, none, "normal memory, inner write-through non-transient, allocate r", none),
                    (0b1011, none, "normal memory, inner write-through non-transient, allocate rw", none),
                    (0b1100, none, "normal memory, inner write-back non-transient", none),
                    (0b1101, none, "normal memory, inner write-back non-transient, allocate w", none),
                    (0b1110, none, "normal memory, inner write-back non-transient, allocate r", none),
                    (0b1111, none, "normal memory, inner write-back non-transient, allocate rw", none),
                ], "device attributes. specifies the memory attributes for device.")
```
here should be "Normal attributes, "  not "Device attribute".

and there should be upcase for
```
 fieldbitfieldenum(f"Inner {num}", offset + 0, 4, [
                    (0b0000, none, "unpredictable", none),
                    (0b0001, none, "Normal memory, inner write-through transient, allocate w", none),
...
```

pengi commented:
You're right that the description is wrong. Should be "Inner attributes. Specifies the Inner memory attributes." according to the reference manual for ARMv8-M. Been trying to copy that where possible.

Been copying the case from that reference manual too, where it keeps "UNPREDICTABLE" in uppercase too. But they have uppercase as references for links to their glossary, it seems. So changing to just "Unpredictable"

I'll push those changes to master. Can create a new version if everything works a bit later again, to give time to test it out a bit more before.
